Added OOL for not having Flash

KamekoSkye opened the pull request:
Added all possible cases where you can access a location or check when you are only lacking flash. This now has them all show up as OOL unless you have flash. For some, this was a simple as changing `$canflash` to `[canflash]`, for others it required adding a possible access requirement needing access to Cerulean and being able to cut. For example, for Lavender Town:
```
"access_rules": [
    "$lavender",
    "$cerulean,$cancut,[canflash]"
]
```
For Fuchsia City:
```
"access_rules": [
    "$fuchsia",
    "$cerulean,$cancut,[canflash],$cansurf"
]
```

This had to be done for all possible ways to a check, since changing logic.lua in the scrips folder would make it think you can do it in logic when you can't, and this therefore was the cleanest way I could think of to add this in.

KamekoSkye committed:
Added OOL for not having Flash

Added all possible cases where you can access a location or check when you are only lacking flash. This now has them all show up as OOL unless you have flash.
For some, this was a simple as changing $canflash to [canflash], for others it required adding a possible access requirement needing access to Cerulean and being able to cut instead of access to a different region

Mindez commented:
As a suggestion - there's a lot of repeated logic; I appreciate that you can't put it in the $fuchsia/$lavender/$celadon check, but it might look cleaner if we have new flags for eg. $fuchsianoflash $lavendernoflash and $celadonnoflash? There'd be a lot less `$cerulean,$cancut,[canflash]` which is a lot more cumbersome to read.

I think there might be some missing logic as well around some of the Fuchsia checks: If you have $cerulean, $cancut, [canflash], $pokeflute, $bike then you can get to Fuchsia with out of logic flash via Rock Tunnel then Cycling Road without surf, if I'm thinking it through correctly. Which I think highlights more the need for a flag to cover the whole out of logic region 😅

KamekoSkye commented:
I assume to create $lavendernoflash etc I would need to add it as a function in logic.lua and then reference it in locations.json? If so, I'll do that and help clean it up (I've never coded using lua, but can work it out with what is already done)

As for having `$cerulean, $cancut, [canflash], pokeflute, bike`, if you have pokeflute you can get to Fuchsia in logic unless extra strength boulders are in place, which is likely the only case I didn't test.

KamekoSkye committed:
Updated OOL tracking

As suggested by Mindez, I have added functions to allow access without flash and made it logically require (but not hard require) flash, showing up as yellow like with Saving Mr. Fuji

KamekoSkye committed:
Last fix as I missed a few things

KamekoSkye commented:
Made the suggested changes. Will test in an AP multiworld now but should work as I haven't deleted the original function, just added new functions that are the same but lack needing flash

KamekoSkye commented:
one needs to be fixed, sorry

KamekoSkye committed:
Fixed OOL bugs

KamekoSkye committed:
Final bugfix, sorry about all that

j-imbo commented:
Now you see why I didn't do it originally. Appreciate all the work here! If you could go through and change all the `[canflash]` into `[$canflash]` I'll merge it. (I know it doesn't actually affect logic due to the original logic getting resolved, but something about it is messing with me.)

KamekoSkye committed:
Changed [canflash] to [$canflash]

KamekoSkye commented:
Done
